pinot-common/src/main/java/com/linkedin/pinot/common/data/DateTimeFieldSpec.java

```diff
@@ -0,0 +1,222 @@
+   */
+  public DateTimeFieldSpec(@Nonnull String name, @Nonnull DataType dataType, @Nonnull String format,
+      @Nonnull String granularity) {
+    super(name, dataType, true);
+    check(name, dataType, format, granularity);
+
+    _format = format;
+    _granularity = granularity;
+  }
+
+
+  /**
+   * @param name
+   * @param dataType
+   * @param format - size:timeunit:timeformat eg: 1:MILLISECONDS:EPOCH, 5:MINUTES:EPOCH, 1:DAYS:SIMPLE_DATE_FORMAT:yyyyMMdd
+   * @param granularity - size:timeunit eg: 5:MINUTES
+   * @param dateTimeType
+   */
+  public DateTimeFieldSpec(@Nonnull String name, @Nonnull DataType dataType, @Nonnull String format,
+      @Nonnull String granularity, DateTimeType dateTimeType) {
+    this(name, dataType, format, granularity);
+    _dateTimeType = dateTimeType;
+  }
+
+  private void check(String name, DataType dataType, String format, String granularity) {
+    Preconditions.checkNotNull(name);
+    Preconditions.checkNotNull(dataType);
+
+    Preconditions.checkNotNull(format);
+    String[] formatTokens = format.split(COLON_SEPARATOR);
+    Preconditions.checkArgument(formatTokens.length == 3 || formatTokens.length == 4,
+        "format must be of format size:timeunit:timeformat(:pattern)");
+    Preconditions.checkArgument(formatTokens[0].matches(NUMBER_REGEX)
+        && EnumUtils.isValidEnum(TimeUnit.class, formatTokens[1]), "format must be of format [0-9]+:<TimeUnit>:<TimeFormat>(:pattern)");
+    if (formatTokens.length == 3) {
+      Preconditions.checkArgument(formatTokens[2].equals(TimeFormat.EPOCH.toString()),
+          "format must be of format [0-9]+:<TimeUnit>:EPOCH or [0-9]+:<TimeUnit>:SIMPLE_DATE_FORMAT:<format>");
+    } else {
+      Preconditions.checkArgument(formatTokens[2].equals(TimeFormat.SIMPLE_DATE_FORMAT.toString()),
+          "format must be of format [0-9]+:<TimeUnit>:EPOCH or [0-9]+:<TimeUnit>:SIMPLE_DATE_FORMAT:<format>");
+    }
+
+    Preconditions.checkNotNull(granularity);
+    String[] granularityTokens = granularity.split(COLON_SEPARATOR);
+    Preconditions.checkArgument(granularityTokens.length == 2, "granularity must be of format size:timeunit");
+    Preconditions.checkArgument(granularityTokens[0].matches(NUMBER_REGEX)
+        && EnumUtils.isValidEnum(TimeUnit.class, granularityTokens[1]), "granularity must be of format [0-9]+:<TimeUnit>");
+
+  }
+
+
+  @JsonIgnore
+  @Nonnull
+  @Override
+  public FieldType getFieldType() {
+    return FieldType.DATE_TIME;
+  }
+
+  @Nonnull
+  public String getFormat() {
+    return _format;
+  }
+
+  public void setFormat(String format) {
+    _format = format;
+  }
+
+  @Nonnull
+  public String getGranularity() {
+    return _granularity;
+  }
+
+  public void setGranularity(String granularity) {
+    _granularity = granularity;
+  }
+
+  @Nullable
+  public DateTimeType getDateTimeType() {
+    return _dateTimeType;
+  }
+
+  public void setDateTimeType(DateTimeType dateTimeType) {
```

Comment:
check which of these setters we actually need. If we can avoid individual setters that is preferred, but I suspect the json parsers need the setters? At the minimum add an annottion & javadoc so that nobody uses the setters. Perhaps @VisibleForTesting will work?

Reply:
Tried to follow what was done in the rest of the setters of the schema, which is either a comment saying "// Required by JSON de-serializer. DO NOT REMOVE." or no comment/annotation at all. Not sure if  @VisibleForTesting  makes sense here. Will add comments/javadocs

Reply:
@VisibleForTesting does not seem to disallow usage, even from IDE. Added javadoc, stating these should not be used.